File: patterns/go/kadane_algorithm.go
package main

import "math"

func maxSubArray(nums []int) int {
	currentSum := nums[0]
	maxSum := nums[0]

	for i := 1; i < len(nums); i++ {
		currentSum = int(math.Max(float64(nums[i]), float64(currentSum+nums[i])))
		maxSum = int(math.Max(float64(maxSum), float64(currentSum)))
	}

	return maxSum
}

File: patterns/c++/KadaneAlgorithm.cpp
#include <vector>
#include <algorithm> // For std::max

class KadaneAlgorithm {
public:
    int maxSubArray(std::vector<int>& nums) {
        int currentSum = nums[0]; // Start with the first element
        int maxSum = nums[0];     // Initialize maxSum with the first element

        // Traverse the array from the second element
        for (size_t i = 1; i < nums.size(); i++) {
            // If currentSum is negative, reset to current element
            currentSum = std::max(nums[i], currentSum + nums[i]);
            // Update maxSum if currentSum is greater
            maxSum = std::max(maxSum, currentSum);
        }
        return maxSum;
    }
};